File: SDF.CSharp/Operations.cs
using System;
using System.Numerics;

namespace SDF;

/// <summary>
/// Boolean operations and transformations for SDFs
/// Operations for transforming and combining SDFs
/// </summary>
public static class Operations
{
    /// <summary>
    /// Union of two SDFs (minimum)
    /// </summary>
    public static SDF3 Union(SDF3 a, SDF3 b, double? k = null)
    {
        k ??= b.SmoothingK ?? a.SmoothingK;

        if (k.HasValue && k.Value > 0)
        {
            return SmoothUnion(a, b, k.Value);
        }

        return new SDF3(points =>
        {
            var da = a.Evaluate(points);
            var db = b.Evaluate(points);
            var result = new double[points.Length];
            for (int i = 0; i < points.Length; i++)
            {
                result[i] = Math.Min(da[i], db[i]);
            }
            return result;
    /// Translate (move) an SDF
    /// </summary>
    public static SDF3 Translate(this SDF3 sdf, Vector3 offset)
    {
        return new SDF3(points =>
        {
            var translatedPoints = new Vector3[points.Length];
            for (int i = 0; i < points.Length; i++)
            {
                translatedPoints[i] = points[i] - offset;
            }
            return sdf.Evaluate(translatedPoints);
        });
    }

    /// <summary>
    /// Scale an SDF uniformly
    /// </summary>
    public static SDF3 Scale(this SDF3 sdf, double factor)
    {
        return new SDF3(points =>
        {
            var scaledPoints = new Vector3[points.Length];
            for (int i = 0; i < points.Length; i++)
            {
                scaledPoints[i] = points[i] / (float)factor;
            }
            var values = sdf.Evaluate(scaledPoints);
            for (int i = 0; i < values.Length; i++)
            {
                values[i] *= factor;
            }
            return values;
        });
    }

    /// <summary>
    /// Smooth union of two SDFs
    /// </summary>
    public static SDF3 SmoothUnion(SDF3 a, SDF3 b, double k)
    {
        if (k <= 0)
            throw new ArgumentException("Smoothing factor k must be positive", nameof(k));

        return new SDF3(points =>
        {
            var da = a.Evaluate(points);
            var db = b.Evaluate(points);
            var result = new double[points.Length];
            for (int i = 0; i < points.Length; i++)
            {
                var h = Math.Max(k - Math.Abs(da[i] - db[i]), 0.0) / k;
                result[i] = Math.Min(da[i], db[i]) - h * h * k * 0.25;
            }
            return result;
    /// Rotate an SDF around an axis
    /// </summary>
    public static SDF3 Rotate(this SDF3 sdf, double angle, Vector3? axis = null)
    {
        var rotationAxis = axis ?? Constants.Z;
        var q = Quaternion.CreateFromAxisAngle(rotationAxis, (float)angle);
        var qInv = Quaternion.Conjugate(q);
        
        return new SDF3(points =>
        {
            var rotatedPoints = new Vector3[points.Length];
            for (int i = 0; i < points.Length; i++)
            {
                rotatedPoints[i] = Vector3.Transform(points[i], qInv);
            }
            return sdf.Evaluate(rotatedPoints);
        });
    }

    /// <summary>
    /// Difference of two SDFs (subtraction)
    /// </summary>
    public static SDF3 Difference(SDF3 a, SDF3 b, double? k = null)
    {
        k ??= b.SmoothingK ?? a.SmoothingK;

        if (k.HasValue && k.Value > 0)
        {
            return SmoothDifference(a, b, k.Value);
        }

        return new SDF3(points =>
        {
            var da = a.Evaluate(points);
            var db = b.Evaluate(points);
            var result = new double[points.Length];
            for (int i = 0; i < points.Length; i++)
            {
                result[i] = Math.Max(da[i], -db[i]);
    /// Orient an SDF to point in a specific direction
    /// </summary>
    public static SDF3 Orient(this SDF3 sdf, Vector3 direction)
    {
        var from = Constants.Z;
        var to = Vector3.Normalize(direction);
        
        // Calculate rotation needed to go from Z to the target direction
        var axis = Vector3.Cross(from, to);
        if (axis.LengthSquared() < 1e-6)
        {
            // Vectors are parallel or anti-parallel
            return Vector3.Dot(from, to) > 0 ? sdf : sdf.Rotate(Math.PI, Constants.X);
        }
        
        axis = Vector3.Normalize(axis);
        var angle = Math.Acos(Vector3.Dot(from, to));
        
        return sdf.Rotate(angle, axis);
    }

    // Boolean Operations
    
    /// <summary>
    /// Union of two SDFs
    /// </summary>
    public static SDF3 Union(SDF3 a, SDF3 b, double? k = null)
    {
        k ??= b.GetK();
        
        if (k.HasValue && k.Value > 0)
        {
            return new SDF3(points =>
            {
                var va = a.Evaluate(points);
                var vb = b.Evaluate(points);
                var result = new double[points.Length];
                
                for (int i = 0; i < points.Length; i++)
                {
                    result[i] = SmoothMin(va[i], vb[i], k.Value);
                }
                return result;
            });
        }
        
        return new SDF3(points =>
        {
            var va = a.Evaluate(points);
            var vb = b.Evaluate(points);
            var result = new double[points.Length];
            
            for (int i = 0; i < points.Length; i++)
            {
                result[i] = Math.Min(va[i], vb[i]);
            }
            return result;
        });
    }

    /// <summary>
    /// Smooth difference of two SDFs
    /// </summary>
    public static SDF3 SmoothDifference(SDF3 a, SDF3 b, double k)
    {
        if (k <= 0)
            throw new ArgumentException("Smoothing factor k must be positive", nameof(k));

        return new SDF3(points =>
        {
            var da = a.Evaluate(points);
            var db = b.Evaluate(points);
            var result = new double[points.Length];
            for (int i = 0; i < points.Length; i++)
            {
                var h = Math.Max(k - Math.Abs(da[i] + db[i]), 0.0) / k;
                result[i] = Math.Max(da[i], -db[i]) + h * h * k * 0.25;
    /// Difference of two SDFs (a - b)
    /// </summary>
    public static SDF3 Difference(SDF3 a, SDF3 b, double? k = null)
    {
        k ??= b.GetK();
        
        if (k.HasValue && k.Value > 0)
        {
            return new SDF3(points =>
            {
                var va = a.Evaluate(points);
                var vb = b.Evaluate(points);
                var result = new double[points.Length];
                
                for (int i = 0; i < points.Length; i++)
                {
                    result[i] = SmoothMax(va[i], -vb[i], k.Value);
                }
                return result;
            });
        }
        
        return new SDF3(points =>
        {
            var va = a.Evaluate(points);
            var vb = b.Evaluate(points);
            var result = new double[points.Length];
            
            for (int i = 0; i < points.Length; i++)
            {
                result[i] = Math.Max(va[i], -vb[i]);
            }
            return result;
        });
    }

    /// <summary>
    /// Intersection of two SDFs (maximum)
    /// </summary>
    public static SDF3 Intersection(SDF3 a, SDF3 b, double? k = null)
    {
        k ??= b.SmoothingK ?? a.SmoothingK;

        if (k.HasValue && k.Value > 0)
        {
            return SmoothIntersection(a, b, k.Value);
        }

        return new SDF3(points =>
        {
            var da = a.Evaluate(points);
            var db = b.Evaluate(points);
            var result = new double[points.Length];
            for (int i = 0; i < points.Length; i++)
            {
                result[i] = Math.Max(da[i], db[i]);
    /// Intersection of two SDFs
    /// </summary>
    public static SDF3 Intersection(SDF3 a, SDF3 b, double? k = null)
    {
        k ??= b.GetK();
        
        if (k.HasValue && k.Value > 0)
        {
            return new SDF3(points =>
            {
                var va = a.Evaluate(points);
                var vb = b.Evaluate(points);
                var result = new double[points.Length];
                
                for (int i = 0; i < points.Length; i++)
                {
                    result[i] = SmoothMax(va[i], vb[i], k.Value);
                }
                return result;
            });
        }
        
        return new SDF3(points =>
        {
            var va = a.Evaluate(points);
            var vb = b.Evaluate(points);
            var result = new double[points.Length];
            
            for (int i = 0; i < points.Length; i++)
            {
                result[i] = Math.Max(va[i], vb[i]);
            }
            return result;
        });
    }

    /// <summary>
    /// Smooth intersection of two SDFs
    /// </summary>
    public static SDF3 SmoothIntersection(SDF3 a, SDF3 b, double k)
    {
        if (k <= 0)
            throw new ArgumentException("Smoothing factor k must be positive", nameof(k));

        return new SDF3(points =>
        {
            var da = a.Evaluate(points);
            var db = b.Evaluate(points);
            var result = new double[points.Length];
            for (int i = 0; i < points.Length; i++)
            {
                var h = Math.Max(k - Math.Abs(da[i] - db[i]), 0.0) / k;
                result[i] = Math.Max(da[i], db[i]) + h * h * k * 0.25;
            }
            return result;
    // Helper functions for smooth operations
    
    private static double SmoothMin(double a, double b, double k)
    {
        var h = Math.Max(k - Math.Abs(a - b), 0.0) / k;
        return Math.Min(a, b) - h * h * k * 0.25;
    }

    private static double SmoothMax(double a, double b, double k)
    {
        return -SmoothMin(-a, -b, k);
    }

    // Deformation Operations

    /// <summary>
    /// Twist an SDF around the Z axis
    /// </summary>
    public static SDF3 Twist(this SDF3 sdf, double k)
    {
        return new SDF3(points =>
        {
            var twistedPoints = new Vector3[points.Length];
            for (int i = 0; i < points.Length; i++)
            {
                var p = points[i];
                var c = Math.Cos(k * p.Z);
                var s = Math.Sin(k * p.Z);
                twistedPoints[i] = new Vector3(
                    (float)(c * p.X - s * p.Y),
                    (float)(s * p.X + c * p.Y),
                    p.Z
                );
            }
            return sdf.Evaluate(twistedPoints);
        });
    }

    /// <summary>
    /// Translate an SDF by an offset
    /// </summary>
    public static SDF3 Translate(this SDF3 sdf, Vector3 offset)
    {
        return new SDF3(points =>
        {
            var translated = new Vector3[points.Length];
            for (int i = 0; i < points.Length; i++)
            {
                translated[i] = points[i] - offset;
            }
            return sdf.Evaluate(translated);
    /// Bend an SDF
    /// </summary>
    public static SDF3 Bend(this SDF3 sdf, double k)
    {
        return new SDF3(points =>
        {
            var bentPoints = new Vector3[points.Length];
            for (int i = 0; i < points.Length; i++)
            {
                var p = points[i];
                var c = Math.Cos(k * p.X);
                var s = Math.Sin(k * p.X);
                bentPoints[i] = new Vector3(
                    (float)(c * p.X - s * p.Y),
                    (float)(s * p.X + c * p.Y),
                    p.Z
                );
            }
            return sdf.Evaluate(bentPoints);
        });
    }

    /// <summary>
    /// Scale an SDF uniformly
    /// </summary>
    public static SDF3 Scale(this SDF3 sdf, double factor)
    {
        return new SDF3(points =>
        {
            var scaled = new Vector3[points.Length];
            for (int i = 0; i < points.Length; i++)
            {
                scaled[i] = points[i] / factor;
            }
            var result = sdf.Evaluate(scaled);
            for (int i = 0; i < result.Length; i++)
            {
                result[i] *= factor;
            }
            return result;
    /// Elongate an SDF along each axis
    /// </summary>
    public static SDF3 Elongate(this SDF3 sdf, Vector3 size)
    {
        return new SDF3(points =>
        {
            var elongatedPoints = new Vector3[points.Length];
            for (int i = 0; i < points.Length; i++)
            {
                var p = points[i];
                elongatedPoints[i] = new Vector3(
                    p.X - Math.Clamp(p.X, -size.X, size.X),
                    p.Y - Math.Clamp(p.Y, -size.Y, size.Y),
                    p.Z - Math.Clamp(p.Z, -size.Z, size.Z)
                );
            }
            return sdf.Evaluate(elongatedPoints);
        });
    }

    /// <summary>
    /// Dilate (expand) an SDF
    /// </summary>
    public static SDF3 Dilate(this SDF3 sdf, double r)
    {
        return new SDF3(points =>
        {
            var values = sdf.Evaluate(points);
            for (int i = 0; i < values.Length; i++)
            {
                values[i] -= r;
            }
            return values;
        });
    }

    /// <summary>
    /// Rotate an SDF around an axis
    /// </summary>
    public static SDF3 Rotate(this SDF3 sdf, double angle, Vector3? axis = null)
    {
        var a = (axis ?? Vector3.UnitZ).Normalize();
        var s = Math.Sin(-angle);
        var c = Math.Cos(-angle);
        var t = 1.0 - c;

        // Rotation matrix
        var m00 = t * a.X * a.X + c;
        var m01 = t * a.X * a.Y - s * a.Z;
        var m02 = t * a.X * a.Z + s * a.Y;
        var m10 = t * a.X * a.Y + s * a.Z;
        var m11 = t * a.Y * a.Y + c;
        var m12 = t * a.Y * a.Z - s * a.X;
        var m20 = t * a.X * a.Z - s * a.Y;
        var m21 = t * a.Y * a.Z + s * a.X;
        var m22 = t * a.Z * a.Z + c;

        return new SDF3(points =>
        {
            var rotated = new Vector3[points.Length];
            for (int i = 0; i < points.Length; i++)
            {
                var p = points[i];
                rotated[i] = new Vector3(
                    m00 * p.X + m01 * p.Y + m02 * p.Z,
                    m10 * p.X + m11 * p.Y + m12 * p.Z,
                    m20 * p.X + m21 * p.Y + m22 * p.Z
                );
            }
            return sdf.Evaluate(rotated);
    /// Erode (shrink) an SDF
    /// </summary>
    public static SDF3 Erode(this SDF3 sdf, double r)
    {
        return new SDF3(points =>
        {
            var values = sdf.Evaluate(points);
            for (int i = 0; i < values.Length; i++)
            {
                values[i] += r;
            }
            return values;
        });
    }

    /// <summary>
    /// Orient an SDF to point in a specific direction (rotates from Z-axis)
    /// </summary>
    public static SDF3 Orient(this SDF3 sdf, Vector3 direction)
    {
        var d = direction.Normalize();
        var z = Vector3.UnitZ;

        // If direction is already aligned with Z, no rotation needed
        if (Math.Abs(Vector3.Dot(d, z) - 1.0) < 1e-6)
        {
            return sdf;
        }

        // If direction is opposite to Z, rotate 180 degrees around X
        if (Math.Abs(Vector3.Dot(d, z) + 1.0) < 1e-6)
        {
            return sdf.Rotate(Math.PI, Vector3.UnitX);
        }

        // Find rotation axis and angle
        var axis = Vector3.Cross(z, d);
        var angle = Math.Acos(Vector3.Dot(z, d));

        return sdf.Rotate(angle, axis);
    /// Create a shell of specified thickness
    /// </summary>
    public static SDF3 Shell(this SDF3 sdf, double thickness)
    {
        return new SDF3(points =>
        {
            var values = sdf.Evaluate(points);
            for (int i = 0; i < values.Length; i++)
            {
                values[i] = Math.Abs(values[i]) - thickness;
            }
            return values;
        });
    }

    /// <summary>
    /// Repeat an SDF with specified spacing
    /// </summary>
    public static SDF3 Repeat(this SDF3 sdf, Vector3 spacing, Vector3? count = null)
    {
        if (count.HasValue)
        {
            // Finite repetition
            var c = count.Value;
            return new SDF3(points =>
            {
                var repeatedPoints = new Vector3[points.Length];
                for (int i = 0; i < points.Length; i++)
                {
                    var p = points[i];
                    repeatedPoints[i] = new Vector3(
                        p.X - spacing.X * (float)Math.Clamp(Math.Round(p.X / spacing.X), -c.X, c.X),
                        p.Y - spacing.Y * (float)Math.Clamp(Math.Round(p.Y / spacing.Y), -c.Y, c.Y),
                        p.Z - spacing.Z * (float)Math.Clamp(Math.Round(p.Z / spacing.Z), -c.Z, c.Z)
                    );
                }
                return sdf.Evaluate(repeatedPoints);
            });
        }
        else
        {
            // Infinite repetition
            return new SDF3(points =>
            {
                var repeatedPoints = new Vector3[points.Length];
                for (int i = 0; i < points.Length; i++)
                {
                    var p = points[i];
                    repeatedPoints[i] = new Vector3(
                        p.X - spacing.X * (float)Math.Round(p.X / spacing.X),
                        p.Y - spacing.Y * (float)Math.Round(p.Y / spacing.Y),
                        p.Z - spacing.Z * (float)Math.Round(p.Z / spacing.Z)
                    );
                }
                return sdf.Evaluate(repeatedPoints);
            });
        }
    }
}

File: SDF.Cpp/src/Operations.cpp
#include "sdf/Operations.h"
#include "sdf/Constants.h"
#include <cmath>
#include <algorithm>

namespace sdf {

// Helper for clamp (C++17 compatible)
template<typename T>
static T clamp(T value, T min, T max) {
    return std::max(min, std::min(value, max));
}

// Boolean operations
SDF3 unionOp(const SDF3& a, const SDF3& b, double k) {
    if (k > 0.0) {
        return a.k(k) | b.k(k);
    }
    return a | b;
}

SDF3 intersection(const SDF3& a, const SDF3& b, double k) {
    if (k > 0.0) {
        return a.k(k) & b.k(k);
    }
    return a & b;
}

SDF3 difference(const SDF3& a, const SDF3& b, double k) {
    if (k > 0.0) {
        return a.k(k) - b.k(k);
    }
    return a - b;
}

// Translate
SDF3 translate(const SDF3& sdf, const Vector3& offset) {
    auto func = sdf.getFunction();
    return SDF3([func, offset](const std::vector<Vector3>& points) {
        std::vector<Vector3> translated;
        translated.reserve(points.size());
        for (const auto& p : points) {
            translated.push_back(p - offset);
        }
        return func(translated);
    });
}

// Scale
SDF3 scale(const SDF3& sdf, double factor) {
    auto func = sdf.getFunction();
    return SDF3([func, factor](const std::vector<Vector3>& points) {
        std::vector<Vector3> scaled;
        scaled.reserve(points.size());
        for (const auto& p : points) {
            scaled.push_back(p / factor);
        }
        auto result = func(scaled);
        for (auto& r : result) {
            r *= factor;
        }
        return result;
    });
}

SDF3 scale(const SDF3& sdf, const Vector3& factor) {
    auto func = sdf.getFunction();
    return SDF3([func, factor](const std::vector<Vector3>& points) {
        std::vector<Vector3> scaled;
        scaled.reserve(points.size());
        for (const auto& p : points) {
            scaled.push_back(p / factor);
        }
        return func(scaled);
    });
}

// Rotate
SDF3 rotate(const SDF3& sdf, double angle, const Vector3& axis) {
    Vector3 a = axis.normalized();
    double c = std::cos(angle);
    double s = std::sin(angle);
    double t = 1.0 - c;
    
    // Rotation matrix (transposed for inverse rotation)
    auto rotatePoint = [a, c, s, t](const Vector3& p) {
        double x = p.x * (t * a.x * a.x + c) + 
                   p.y * (t * a.x * a.y + s * a.z) + 
                   p.z * (t * a.x * a.z - s * a.y);
        double y = p.x * (t * a.x * a.y - s * a.z) + 
                   p.y * (t * a.y * a.y + c) + 
                   p.z * (t * a.y * a.z + s * a.x);
        double z = p.x * (t * a.x * a.z + s * a.y) + 
                   p.y * (t * a.y * a.z - s * a.x) + 
                   p.z * (t * a.z * a.z + c);
        return Vector3(x, y, z);
    };
    
    auto func = sdf.getFunction();
    return SDF3([func, rotatePoint](const std::vector<Vector3>& points) {
        std::vector<Vector3> rotated;
        rotated.reserve(points.size());
        for (const auto& p : points) {
            rotated.push_back(rotatePoint(p));
        }
        return func(rotated);
    });
}

// Orient
SDF3 orient(const SDF3& sdf, const Vector3& targetAxis) {
    Vector3 src = Z;
    Vector3 dst = targetAxis.normalized();
    
    // If already aligned, no rotation needed
    if ((src - dst).lengthSquared() < 1e-10) {
        return sdf;
    }
    
    // If opposite direction, rotate 180 degrees around perpendicular axis
    if ((src + dst).lengthSquared() < 1e-10) {
        Vector3 perp = (std::abs(src.x) < 0.9) ? Vector3(1, 0, 0) : Vector3(0, 1, 0);
        return rotate(sdf, PI, perp);
    }
    
    // Calculate rotation axis and angle
    Vector3 axis = src.cross(dst).normalized();
    double angle = std::acos(clamp(src.dot(dst), -1.0, 1.0));
    
    return rotate(sdf, angle, axis);
}

// Twist
SDF3 twist(const SDF3& sdf, double k) {
    auto func = sdf.getFunction();
    return SDF3([func, k](const std::vector<Vector3>& points) {
        std::vector<Vector3> twisted;
        twisted.reserve(points.size());
        for (const auto& p : points) {
            double c = std::cos(k * p.z);
            double s = std::sin(k * p.z);
            twisted.push_back(Vector3(
                c * p.x - s * p.y,
                s * p.x + c * p.y,
                p.z
            ));
        }
        return func(twisted);
    });
}

// Bend
SDF3 bend(const SDF3& sdf, double k) {
    auto func = sdf.getFunction();
    return SDF3([func, k](const std::vector<Vector3>& points) {
        std::vector<Vector3> bent;
        bent.reserve(points.size());
        for (const auto& p : points) {
            double c = std::cos(k * p.x);
            double s = std::sin(k * p.x);
            bent.push_back(Vector3(
                c * p.x - s * p.y,
                s * p.x + c * p.y,
                p.z
            ));
        }
        return func(bent);
    });
}

// Elongate
SDF3 elongate(const SDF3& sdf, const Vector3& h) {
    auto func = sdf.getFunction();
    return SDF3([func, h](const std::vector<Vector3>& points) {
        std::vector<Vector3> elongated;
        elongated.reserve(points.size());
        for (const auto& p : points) {
            Vector3 q = Vector3::abs(p) - h;
            elongated.push_back(Vector3(
                std::max(q.x, 0.0) * (p.x >= 0 ? 1.0 : -1.0),
                std::max(q.y, 0.0) * (p.y >= 0 ? 1.0 : -1.0),
                std::max(q.z, 0.0) * (p.z >= 0 ? 1.0 : -1.0)
            ));
        }
        auto result = func(elongated);
        
        // Adjust for exterior distance
        for (size_t i = 0; i < points.size(); ++i) {
            Vector3 q = Vector3::abs(points[i]) - h;
            Vector3 maxQ = Vector3::max(q, Vector3(0, 0, 0));
            result[i] += maxQ.length();
        }
        return result;
    });
}

// Dilate
SDF3 dilate(const SDF3& sdf, double r) {
    auto func = sdf.getFunction();
    return SDF3([func, r](const std::vector<Vector3>& points) {
        auto result = func(points);
        for (auto& d : result) {
            d -= r;
        }
        return result;
    });
}

// Erode
SDF3 erode(const SDF3& sdf, double r) {
    return dilate(sdf, -r);
}

// Shell
SDF3 shell(const SDF3& sdf, double thickness) {
    auto func = sdf.getFunction();
    return SDF3([func, thickness](const std::vector<Vector3>& points) {
        auto result = func(points);
        for (auto& d : result) {
            d = std::abs(d) - thickness;
        }
        return result;
    });
}

// Repeat
SDF3 repeatOp(const SDF3& sdf, const Vector3& spacing, const Vector3& count) {
    auto func = sdf.getFunction();
    return SDF3([func, spacing, count](const std::vector<Vector3>& points) {
        std::vector<Vector3> repeated;
        repeated.reserve(points.size());
        
        for (const auto& p : points) {
            Vector3 q = p;
            
            // Apply modulo for each axis if count is not infinite
            if (count.x < 1e8) {
                double cx = std::round(p.x / spacing.x);
                cx = clamp(cx, -count.x, count.x);
                q.x = p.x - cx * spacing.x;
            } else {
                q.x = p.x - std::round(p.x / spacing.x) * spacing.x;
            }
            
            if (count.y < 1e8) {
                double cy = std::round(p.y / spacing.y);
                cy = clamp(cy, -count.y, count.y);
                q.y = p.y - cy * spacing.y;
            } else {
                q.y = p.y - std::round(p.y / spacing.y) * spacing.y;
            }
            
            if (count.z < 1e8) {
                double cz = std::round(p.z / spacing.z);
                cz = clamp(cz, -count.z, count.z);
                q.z = p.z - cz * spacing.z;
            } else {
                q.z = p.z - std::round(p.z / spacing.z) * spacing.z;
            }
            
            repeated.push_back(q);
        }
        
        return func(repeated);
    });
}

// Blend
SDF3 blend(const SDF3& a, const SDF3& b, double k) {
    auto funcA = a.getFunction();
    auto funcB = b.getFunction();
    
    return SDF3([funcA, funcB, k](const std::vector<Vector3>& points) {
        auto ra = funcA(points);
        auto rb = funcB(points);
        std::vector<double> result;
        result.reserve(points.size());
        
        for (size_t i = 0; i < points.size(); ++i) {
            result.push_back(ra[i] * (1.0 - k) + rb[i] * k);
        }
        
        return result;
    });
}

// Circular Array
SDF3 circularArray(const SDF3& sdf, int count, double offset) {
    auto func = sdf.getFunction();
    
    return SDF3([func, count, offset](const std::vector<Vector3>& points) {
        std::vector<double> result(points.size(), 1e9);
        
        for (int i = 0; i < count; ++i) {
            double angle = 2.0 * PI * i / count;
            double c = std::cos(angle);
            double s = std::sin(angle);
            
            std::vector<Vector3> transformed;
            transformed.reserve(points.size());
            
            for (const auto& p : points) {
                // Rotate around Z axis and translate
                double x = c * (p.x - offset) - s * p.y + offset;
                double y = s * (p.x - offset) + c * p.y;
                transformed.push_back(Vector3(x, y, p.z));
            }
            
            auto distances = func(transformed);
            for (size_t j = 0; j < points.size(); ++j) {
                result[j] = std::min(result[j], distances[j]);
            }
        }
        
        return result;
    });
}

} // namespace sdf
